File: tool/level_editor/src/utils/LevelBinaryCodec.ts
import {
  ArrowDirection,
  type ArrowModel,
  LevelDifficulty,
  type LevelModel,
  MaskShape,
  type OrphanDotModel,
  OrphanDotType,
  SnakeMechanic,
} from '../models/types';

const MAGIC = [0x4c, 0x56, 0x4c, 0x42]; // 'LVLB'
const VERSION = 2;

// ─── Encoder Class ────────────────────────────────────────────────────────────

class ByteWriter {
  private buffer: number[] = [];

  writeUint8(val: number) {
    this.buffer.push(val & 0xff);
  }

  writeUint16(val: number) {
    this.buffer.push(val & 0xff);
    this.buffer.push((val >> 8) & 0xff); // Little endian
  }

  writeUint32(val: number) {
    this.buffer.push(val & 0xff);
    this.buffer.push((val >> 8) & 0xff);
    this.buffer.push((val >> 16) & 0xff);
    this.buffer.push((val >> 24) & 0xff);
  }

  writeString(str: string) {
    const bytes = new TextEncoder().encode(str);
    this.writeUint8(bytes.length);
    for (let i = 0; i < bytes.length; i++) {
      this.buffer.push(bytes[i]);
    }
  }

  writeBytes(bytes: Uint8Array | number[]) {
    for (let i = 0; i < bytes.length; i++) {
      this.buffer.push(bytes[i]);
    }
  }

  getBytes(): Uint8Array {
    return new Uint8Array(this.buffer);
  }

  get length(): number {
    return this.buffer.length;
  }
}

class ByteReader {
  private view: DataView;
  private offset = 0;

  constructor(private buffer: Uint8Array) {
    this.view = new DataView(
      buffer.buffer,
      buffer.byteOffset,
      buffer.byteLength
    );
  }

  get position(): number {
    return this.offset;
  }

  set position(pos: number) {
    this.offset = pos;
  }

  readUint8(): number {
    const val = this.view.getUint8(this.offset);
    this.offset += 1;
    return val;
  }

  readUint16(): number {
    const val = this.view.getUint16(this.offset, true); // Little endian
    this.offset += 2;
    return val;
  }

  readUint32(): number {
    const val = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return val;
  }

  readString(): string {
    const len = this.readUint8();
    const bytes = this.buffer.subarray(this.offset, this.offset + len);
    this.offset += len;
    return new TextDecoder().decode(bytes);
  }

  readBytes(count: number): Uint8Array {
    const bytes = this.buffer.subarray(this.offset, this.offset + count);
    this.offset += count;
    return bytes;
  }
}

// ─── Encode Levels to Uint8Array ──────────────────────────────────────────────

export function encodeLevelsToBinary(levels: LevelModel[]): Uint8Array {
  const encodedLevels = levels.map((l) => encodeSingleLevel(l));

  const headerSize = 8;
  const indexSize = levels.length * 4;
  let dataSize = 0;
  for (const b of encodedLevels) dataSize += b.length;

  const totalSize = headerSize + indexSize + dataSize;
  const out = new ByteWriter();

  // Header
  for (const b of MAGIC) out.writeUint8(b);
  out.writeUint16(VERSION);
  out.writeUint16(levels.length);

  // Index table
  let dataOffset = 0;
  for (const encoded of encodedLevels) {
    out.writeUint32(dataOffset);
    dataOffset += encoded.length;
  }

  // Data Section
  for (const encoded of encodedLevels) {
    out.writeBytes(encoded);
  }

  return out.getBytes();
}

function encodeSingleLevel(level: LevelModel): Uint8Array {
  const buf = new ByteWriter();

  buf.writeUint16(level.levelNumber);
  buf.writeUint8(level.gridSize);
  buf.writeUint8(level.maskShape ?? MaskShape.fullGrid);
  buf.writeUint8(level.difficulty ?? LevelDifficulty.medium);
  buf.writeString(level.patternName ?? '');
  buf.writeUint16(level.arrows.length);

  for (const arrow of level.arrows) {
    buf.writeUint8(arrow.row);
    buf.writeUint8(arrow.col);

    let dirVal = 0;
    switch (arrow.direction) {
      case ArrowDirection.up:    dirVal = 0; break;
      case ArrowDirection.down:  dirVal = 1; break;
      case ArrowDirection.left:  dirVal = 2; break;
      case ArrowDirection.right: dirVal = 3; break;
    }

    const mechVal = arrow.mechanic & 0x03;
    const patVal = arrow.isPartOfPattern ? 1 : 0;
    const packed = (dirVal & 0x03) | ((mechVal & 0x03) << 2) | ((patVal & 0x01) << 4);
    buf.writeUint8(packed);

    buf.writeUint8(arrow.colorGroup ?? 0xff);

    // Path steps (steps AFTER head)
    const stepCount = Math.max(0, arrow.path.length - 1);
    buf.writeUint8(stepCount);

    for (let i = 0; i < stepCount; i++) {
      const curr = arrow.path[i];
      const next = arrow.path[i + 1];
      const dr = next[0] - curr[0];
      const dc = next[1] - curr[1];

      let stepCode = 3; // right
      if (dr === -1 && dc === 0) stepCode = 0;      // up
      else if (dr === 1 && dc === 0) stepCode = 1;  // down
      else if (dr === 0 && dc === -1) stepCode = 2; // left
      else if (dr === 0 && dc === 1) stepCode = 3;  // right
      buf.writeUint8(stepCode);
    }
  }

  const sol = level.solutionOrder ?? [];
  buf.writeUint16(sol.length);
  for (const idx of sol) buf.writeUint8(idx);

  // Bitmask
  const totalCells = level.gridSize * level.gridSize;
  const bitmaskBytes = Math.ceil(totalCells / 8);
  if (level.maskBitmask && level.maskBitmask.length === bitmaskBytes) {
    buf.writeBytes(level.maskBitmask);
  } else {
    // Fill all 1s (full grid active)
    const fill = new Uint8Array(bitmaskBytes);
    fill.fill(0xff);
    buf.writeBytes(fill);
  }

  // Orphan dots
  const dots = level.orphanDots ?? [];
  buf.writeUint16(dots.length);
  for (const d of dots) {
    buf.writeUint8(d.row);
    buf.writeUint8(d.col);
    let typeCode = 0;
    switch (d.type) {
      case OrphanDotType.up:    typeCode = 0; break;
      case OrphanDotType.down:  typeCode = 1; break;
      case OrphanDotType.left:  typeCode = 2; break;
      case OrphanDotType.right: typeCode = 3; break;
    }
    buf.writeUint8(typeCode);
  }

  return buf.getBytes();
}

// ─── Decode Binary to LevelModel[] ───────────────────────────────────────────

export function decodeBinaryToLevels(data: Uint8Array): LevelModel[] {
  const reader = new ByteReader(data);

  for (let i = 0; i < 4; i++) {
    if (reader.readUint8() !== MAGIC[i]) {
      throw new Error('Invalid magic header — not a LVLB level file');
    }
  }

  const ver = reader.readUint16();
  if (ver !== VERSION) {
    throw new Error(`Unsupported LVLB version: ${ver}`);
  }

  const levelCount = reader.readUint16();
  const offsets: number[] = [];
  for (let i = 0; i < levelCount; i++) {
    offsets.push(reader.readUint32());
  }

  const dataStart = reader.position;
  const levels: LevelModel[] = [];

  for (let i = 0; i < levelCount; i++) {
    reader.position = dataStart + offsets[i];
    levels.push(decodeSingleLevel(reader, i + 1));
  }

  return levels;
}

function decodeSingleLevel(reader: ByteReader, index: number): LevelModel {
  const levelNumber = reader.readUint16();
  const gridSize = reader.readUint8();
  const maskShape = reader.readUint8() as MaskShape;
  const difficulty = reader.readUint8() as LevelDifficulty;
  const patternName = reader.readString();
  const arrowCount = reader.readUint16();

  const arrows: ArrowModel[] = [];

  for (let i = 0; i < arrowCount; i++) {
    const headRow = reader.readUint8();
    const headCol = reader.readUint8();
    const packed = reader.readUint8();

    const dirCode = packed & 0x03;
    const mechCode = (packed >> 2) & 0x03;
    const isPartOfPattern = ((packed >> 4) & 0x01) === 1;

    let dir = ArrowDirection.up;
    if (dirCode === 0) dir = ArrowDirection.up;
    else if (dirCode === 1) dir = ArrowDirection.down;
    else if (dirCode === 2) dir = ArrowDirection.left;
    else if (dirCode === 3) dir = ArrowDirection.right;

    const rawColor = reader.readUint8();
    const colorGroup = rawColor === 0xff ? null : rawColor;
    const stepCount = reader.readUint8();

    const path: [number, number][] = [[headRow, headCol]];
    let currRow = headRow;
    let currCol = headCol;

    for (let s = 0; s < stepCount; s++) {
      const stepCode = reader.readUint8();
      if (stepCode === 0) currRow -= 1;      // up
      else if (stepCode === 1) currRow += 1; // down
      else if (stepCode === 2) currCol -= 1; // left
      else if (stepCode === 3) currCol += 1; // right
      path.push([currRow, currCol]);
    }

    arrows.push({
      id: `arrow_${index}_${i}`,
      row: headRow,
      col: headCol,
      direction: dir,
      isPartOfPattern,
      mechanic: mechCode as SnakeMechanic,
      colorGroup,
      path,
    });
  }

  const solCount = reader.readUint16();
  const solutionOrder: number[] = [];
  for (let i = 0; i < solCount; i++) {
    solutionOrder.push(reader.readUint8());
  }

  const totalCells = gridSize * gridSize;
  const bitmaskBytes = Math.ceil(totalCells / 8);
  const maskBitmask = reader.readBytes(bitmaskBytes);

  const dotCount = reader.readUint16();
  const orphanDots: OrphanDotModel[] = [];

  for (let i = 0; i < dotCount; i++) {
    const row = reader.readUint8();
    const col = reader.readUint8();
    const typeCode = reader.readUint8();
    let type = OrphanDotType.up;
    if (typeCode === 0) type = OrphanDotType.up;
    else if (typeCode === 1) type = OrphanDotType.down;
    else if (typeCode === 2) type = OrphanDotType.left;
    else if (typeCode === 3) type = OrphanDotType.right;

    orphanDots.push({ row, col, type });
  }

  return {
    levelNumber,
    gridSize,
    maskShape,
    difficulty,
    patternName,
    arrows,
    solutionOrder,
    maskBitmask,
    orphanDots,
  };
}

File: lib/data/level_binary_codec.dart
import 'dart:convert';
import 'dart:typed_data';
import 'models/arrow.dart';
import 'models/level.dart';

// ─── Binary Format Spec (v2) ──────────────────────────────────────────────────
//
// HEADER         (8 bytes)
//   4 bytes  magic: 0x4C 0x56 0x4C 0x42  ('LVLB')
//   2 bytes  version: uint16 = 2
//   2 bytes  levelCount: uint16
//
// INDEX TABLE    (levelCount × 4 bytes)
//   uint32   byte offset of each level's data, relative to start of DATA SECTION
//
// DATA SECTION   (variable)
//   For each level:
//     2 bytes  levelNumber (uint16)
//     1 byte   gridSize
//     1 byte   maskShape index
//     1 byte   difficulty index
//     1 byte   patternName length (in bytes)
//     N bytes  patternName (UTF-8)
//     2 bytes  arrowCount (uint16)
//
//     [ARROWS] × arrowCount
//       1 byte  row
//       1 byte  col
//       1 byte  packed: direction(2b) | mechanic(2b) | isPartOfPattern(1b) | reserved(3b)
//       1 byte  colorGroup (0xFF = null)
//       1 byte  pathStepCount  (steps AFTER the head cell)
//       pathStepCount bytes: step direction (0=up,1=down,2=left,3=right)
//
//     2 bytes  solutionOrderCount (uint16)
//     [SOLUTION ORDER] × solutionOrderCount bytes: arrow index (uint8)
//       NOTE: if arrowCount > 255 this would need uint16 — but in practice max is ~200
//
//     ceil(gridSize²/8) bytes: mask bitmask, row-major, MSB first
//
//     2 bytes  orphanDotCount (uint16)
//     [ORPHAN DOTS] × orphanDotCount × 3 bytes:
//       1 byte  row
//       1 byte  col
//       1 byte  type index
//
// ─────────────────────────────────────────────────────────────────────────────

const _kMagic = [0x4C, 0x56, 0x4C, 0x42]; // 'LVLB'
const _kVersion = 2;

// ─── Encoder ─────────────────────────────────────────────────────────────────

/// Encodes a list of [LevelModel] objects into a compact binary [Uint8List].
Uint8List encodeLevels(List<LevelModel> levels) {
  // 1. Encode each level individually, collect byte arrays.
  final encodedLevels = levels.map(_encodeLevel).toList();

  final headerSize = 8;
  final indexSize = levels.length * 4;
  final dataSize = encodedLevels.fold<int>(0, (sum, b) => sum + b.length);
  final totalSize = headerSize + indexSize + dataSize;

  final out = ByteData(totalSize);
  int pos = 0;

  // Header
  for (final b in _kMagic) {
    out.setUint8(pos++, b);
  }
  out.setUint16(pos, _kVersion, Endian.little);
  pos += 2;
  out.setUint16(pos, levels.length, Endian.little);
  pos += 2;

  // Index table — offsets relative to start of data section
  int dataOffset = 0;
  for (final encoded in encodedLevels) {
    out.setUint32(pos, dataOffset, Endian.little);
    pos += 4;
    dataOffset += encoded.length;
  }

  // Data section
  for (final encoded in encodedLevels) {
    for (int i = 0; i < encoded.length; i++) {
      out.setUint8(pos++, encoded[i]);
    }
  }

  return out.buffer.asUint8List();
}

Uint8List _encodeLevel(LevelModel level) {
  final buf = _ByteWriter();

  buf.writeUint16(level.levelNumber);
  buf.writeUint8(level.gridSize);
  buf.writeUint8(level.maskShape.index);
  buf.writeUint8(level.difficulty.index);
  buf.writeString(level.patternName);
  buf.writeUint16(level.arrows.length);

  for (final arrow in level.arrows) {
    buf.writeUint8(arrow.row);
    buf.writeUint8(arrow.col);

    // packed byte: direction(2b) | mechanic(2b) | isPartOfPattern(1b) | 0(3b)
    final packed = (arrow.direction.index & 0x3) |
        ((arrow.mechanic.index & 0x3) << 2) |
        ((arrow.isPartOfPattern ? 1 : 0) << 4);
    buf.writeUint8(packed);

    buf.writeUint8(arrow.colorGroup ?? 0xFF);

    // Path: head is at path[0], each subsequent step encoded as a delta direction.
    final steps = _encodePathSteps(arrow.path);
    buf.writeUint8(steps.length); // max path length ~255 in practice
    for (final s in steps) {
      buf.writeUint8(s);
    }
  }

  // Solution order: stored as arrow indices
  final arrowIdToIndex = <String, int>{};
  for (int i = 0; i < level.arrows.length; i++) {
    arrowIdToIndex[level.arrows[i].id] = i;
  }
  buf.writeUint16(level.solutionOrder.length);
  for (final id in level.solutionOrder) {
    buf.writeUint8(arrowIdToIndex[id] ?? 0);
  }

  // Mask as bitmask (row-major, MSB first within each byte)
  final gs = level.gridSize;
  final maskBitCount = gs * gs;
  final maskByteCount = (maskBitCount + 7) ~/ 8;
  final maskBytes = Uint8List(maskByteCount);
  for (final cell in level.mask) {
    final comma = cell.indexOf(',');
    final r = int.parse(cell.substring(0, comma));
    final c = int.parse(cell.substring(comma + 1));
    final bit = r * gs + c;
    maskBytes[bit >> 3] |= (0x80 >> (bit & 7));
  }
  buf.writeBytes(maskBytes);

  // Orphan dots
  buf.writeUint16(level.orphanDots.length);
  for (final dot in level.orphanDots) {
    buf.writeUint8(dot.row);
    buf.writeUint8(dot.col);
    buf.writeUint8(dot.type.index);
  }

  return buf.toBytes();
}

/// Converts path (list of [row,col] pairs) to delta step directions.
List<int> _encodePathSteps(List<List<int>> path) {
  final steps = <int>[];
  for (int i = 0; i < path.length - 1; i++) {
    final dr = path[i + 1][0] - path[i][0];
    final dc = path[i + 1][1] - path[i][1];
    if (dr == -1) steps.add(0);      // up
    else if (dr == 1) steps.add(1);  // down
    else if (dc == -1) steps.add(2); // left
    else steps.add(3);               // right
  }
  return steps;
}

// ─── Decoder ─────────────────────────────────────────────────────────────────

/// Decodes binary level data produced by [encodeLevels].
/// Provides O(1) random access to any level via an index table.
class LevelBinaryDecoder {
  final ByteData _data;
  final int _levelCount;
  final int _indexTableStart;
  final int _dataStart;

  LevelBinaryDecoder._(
      this._data, this._levelCount, this._indexTableStart, this._dataStart);

  /// Parse the header and construct the decoder. Throws on magic/version mismatch.
  factory LevelBinaryDecoder.fromBytes(Uint8List bytes) {
    final data = ByteData.sublistView(bytes);
    // Validate magic bytes
    for (int i = 0; i < 4; i++) {
      if (data.getUint8(i) != _kMagic[i]) {
        throw const FormatException('Invalid levels.bin: bad magic');
      }
    }
    final version = data.getUint16(4, Endian.little);
    if (version != _kVersion) {
      throw FormatException('Unsupported levels.bin version: $version');
    }
    final levelCount = data.getUint16(6, Endian.little);
    const indexTableStart = 8;
    final dataStart = indexTableStart + levelCount * 4;
    return LevelBinaryDecoder._(data, levelCount, indexTableStart, dataStart);
  }

  int get levelCount => _levelCount;

  /// Decode a level by its 1-based level number. Returns null if out of range.
  LevelModel? decodeLevelByNumber(int levelNumber) {
    final idx = levelNumber - 1;
    if (idx < 0 || idx >= _levelCount) return null;
    return _decodeLevelAtIndex(idx);
  }

  /// Decode all levels eagerly (only for tools, not for device use).
  List<LevelModel> decodeAll() {
    return List.generate(_levelCount, _decodeLevelAtIndex);
  }

  LevelModel _decodeLevelAtIndex(int idx) {
    final offsetInData =
        _data.getUint32(_indexTableStart + idx * 4, Endian.little);
    final reader = _ByteReader(_data, _dataStart + offsetInData);
    return _decodeLevel(reader);
  }

  LevelModel _decodeLevel(_ByteReader r) {
    final levelNumber = r.readUint16();
    final gridSize = r.readUint8();
    final maskShape = MaskShape.values[
        r.readUint8().clamp(0, MaskShape.values.length - 1)];
    final difficulty = Difficulty.values[
        r.readUint8().clamp(0, Difficulty.values.length - 1)];
    final patternName = r.readString();
    final arrowCount = r.readUint16();

    final arrows = <ArrowModel>[];
    for (int i = 0; i < arrowCount; i++) {
      final row = r.readUint8();
      final col = r.readUint8();
      final packed = r.readUint8();
      final direction = ArrowDirection.values[packed & 0x3];
      final mechanic = SnakeMechanic.values[(packed >> 2) & 0x3];
      final isPartOfPattern = ((packed >> 4) & 0x1) == 1;
      final colorGroupRaw = r.readUint8();
      final colorGroup = colorGroupRaw == 0xFF ? null : colorGroupRaw;

      final stepCount = r.readUint8();
      final path = <List<int>>[
        [row, col]
      ];
      for (int s = 0; s < stepCount; s++) {
        final stepDir = r.readUint8();
        final last = path.last;
        switch (stepDir) {
          case 0:
            path.add([last[0] - 1, last[1]]);
            break; // up
          case 1:
            path.add([last[0] + 1, last[1]]);
            break; // down
          case 2:
            path.add([last[0], last[1] - 1]);
            break; // left
          default:
            path.add([last[0], last[1] + 1]);
            break; // right
        }
      }

      arrows.add(ArrowModel(
        id: 'a_${levelNumber}_$i',
        row: row,
        col: col,
        direction: direction,
        state: ArrowState.idle,
        isPartOfPattern: isPartOfPattern,
        mechanic: mechanic,
        colorGroup: colorGroup,
        path: path,
      ));
    }

    // Solution order — stored as arrow index bytes
    final solCount = r.readUint16();
    final solutionOrder = <String>[];
    for (int i = 0; i < solCount; i++) {
      final arrowIdx = r.readUint8();
      solutionOrder.add('a_${levelNumber}_$arrowIdx');
    }

    // Mask bitmask — row-major, MSB first
    final maskBitCount = gridSize * gridSize;
    final maskByteCount = (maskBitCount + 7) ~/ 8;
    final maskBytes = r.readBytes(maskByteCount);
    final mask = <String>{};
    for (int bit = 0; bit < maskBitCount; bit++) {
      if ((maskBytes[bit >> 3] & (0x80 >> (bit & 7))) != 0) {
        final row = bit ~/ gridSize;
        final col = bit % gridSize;
        mask.add('$row,$col');
      }
    }

    // Orphan dots
    final orphanCount = r.readUint16();
    final orphanDots = <OrphanDot>[];
    for (int i = 0; i < orphanCount; i++) {
      final row = r.readUint8();
      final col = r.readUint8();
      final type = OrphanDotType.values[
          r.readUint8().clamp(0, OrphanDotType.values.length - 1)];
      orphanDots.add(OrphanDot(row: row, col: col, type: type));
    }

    return LevelModel(
      levelNumber: levelNumber,
      gridSize: gridSize,
      arrows: arrows,
      patternName: patternName,
      difficulty: difficulty,
      solutionOrder: solutionOrder,
      maskShape: maskShape,
      mask: mask,
      orphanDots: orphanDots,
    );
  }
}

// ─── Internal helpers ─────────────────────────────────────────────────────────

class _ByteWriter {
  final List<int> _bytes = [];

  void writeUint8(int v) => _bytes.add(v & 0xFF);

  void writeUint16(int v) {
    _bytes.add(v & 0xFF);
    _bytes.add((v >> 8) & 0xFF);
  }

  /// Write a short UTF-8 string (length-prefixed, max 255 bytes).
  /// Longer input is truncated to fit, at a valid UTF-8 character boundary
  /// (never splitting a multi-byte sequence) - the 1-byte length prefix
  /// (writeUint8) would otherwise silently wrap for anything over 255
  /// bytes while the full string still got written, corrupting every
  /// field decoded after it.
  void writeString(String s) {
    var encoded = utf8.encode(s);
    if (encoded.length > 255) {
      var cut = 255;
      while (cut > 0 && (encoded[cut] & 0xC0) == 0x80) {
        cut--;
      }
      encoded = encoded.sublist(0, cut);
    }
    writeUint8(encoded.length);
    _bytes.addAll(encoded);
  }

  void writeBytes(Uint8List bytes) => _bytes.addAll(bytes);

  Uint8List toBytes() => Uint8List.fromList(_bytes);
}

class _ByteReader {
  final ByteData _data;
  int _pos;

  _ByteReader(this._data, int start) : _pos = start;

  int readUint8() => _data.getUint8(_pos++);

  int readUint16() {
    final v = _data.getUint16(_pos, Endian.little);
    _pos += 2;
    return v;
  }

  /// Read a length-prefixed UTF-8 string.
  String readString() {
    final len = readUint8();
    final bytes = _data.buffer.asUint8List(_data.offsetInBytes + _pos, len);
    _pos += len;
    return utf8.decode(bytes);
  }

  Uint8List readBytes(int count) {
    final bytes =
        _data.buffer.asUint8List(_data.offsetInBytes + _pos, count);
    _pos += count;
    return bytes;
  }
}
